tiledb/sm/subarray/subarray.cc

```diff
@@ -2181,60 +2181,181 @@ Status Subarray::compute_relevant_fragments(
+  fn_ctx->last_start_coords_ = start_coords;
+  fn_ctx->last_end_coords_ = end_coords;
+
+  // Populate the fragment bytemap for each dimension in parallel.
+  RETURN_NOT_OK(parallel_for(compute_tp, 0, dim_num, [&](const uint32_t d) {
+    return compute_relevant_fragments_for_dim(
+        compute_tp,
+        d,
+        fragment_num,
+        start_coords,
+        end_coords,
+        &fn_ctx->frag_bytemaps_[d]);
+  }));
+
+  // Recalculate relevant fragments.
   relevant_fragments_.reserve(fragment_num);
+  relevant_fragments_.clear();
   for (unsigned f = 0; f < fragment_num; ++f) {
-    if (fn_ctx->frag_bytemap_[f])
+    bool relevant = true;
+    for (uint32_t d = 0; d < dim_num; ++d) {
+      if (fn_ctx->frag_bytemaps_[d][f] == 0) {
+        relevant = false;
+        break;
+      }
+    }
+
+    if (relevant) {
       relevant_fragments_.emplace_back(f);
+    }
   }
-  relevant_fragments_.shrink_to_fit();
 
   return Status::Ok();
 
   STATS_END_TIMER(stats::GlobalStats::TimerType::READ_COMPUTE_RELEVANT_FRAGS)
 }
 
+void Subarray::get_expanded_coordinates(
+    const uint64_t range_idx_start,
+    const uint64_t range_idx_end,
+    std::vector<uint64_t>* const start_coords,
+    std::vector<uint64_t>* const end_coords) const {
+  // Fetch the multi-dimensional coordinates from the
+  // flattened (total order) range indexes.
+  *start_coords = get_range_coords(range_idx_start);
+  *end_coords = get_range_coords(range_idx_end);
+
+  // The rest of this routine is concerned with expanding
+  // the coordinates to ensure that the n-dimensional space
+  // between `start_coords` and `end_coords` encapsulates
+  // all ranges between the total-ordered start/end indexes.
+  // This is non-applicable to one-dimensional arrays.
+  const uint32_t dim_num = array_->array_schema()->dim_num();
+  assert(dim_num > 0);
+  if (dim_num == 1) {
+    return;
+  }
+
+  // This is only applicable to row-major, column-major, or unordered
+  // layouts. We will treat unordered layouts as the cell layout.
+  assert(
+      layout_ == Layout::ROW_MAJOR || layout_ == Layout::COL_MAJOR ||
+      layout_ == Layout::UNORDERED);
+  const Layout coords_layout =
+      (layout_ == Layout::UNORDERED) ? cell_order_ : layout_;
+  assert(
+      coords_layout == Layout::ROW_MAJOR || coords_layout == Layout::COL_MAJOR);
+
+  // If the n-dimensional space exists within the same index on
+  // all but the last dimension, we do not need to expand the
+  // coordinates.
+  bool expand = false;
+  if (coords_layout == Layout::ROW_MAJOR) {
+    for (uint32_t d = 0; d < dim_num - 1; ++d) {
+      if ((*start_coords)[d] != (*end_coords)[d]) {
+        expand = true;
+        break;
+      }
+    }
+  } else {
+    assert(coords_layout == Layout::COL_MAJOR);
+    for (uint32_t d = 1; d < dim_num; ++d) {
+      if ((*start_coords)[d] != (*end_coords)[d]) {
+        expand = true;
+        break;
+      }
+    }
+  }
+
+  // Expand the coordinates if necessary.
+  if (expand) {
```

Comment:
Here you are over-expanding. You need to expand only the "last" dimension.

Consider row-major, cube 4 x 5 x 3, and ranges 0-3. The start coords are (0, 0, 0) and the last coords are (0, 1, 0). This has to be expanded to [0,0] x [0,1] x [0,3], not [0,0] x [0,4] x [0,3].

Reply:
That's a good point: this over-expands for that specific scenario. However, I can provide a counter-point (and have included them in the unit tests), where we need to expand more than the last dimension.

I have determined that it is only necessary to expand the dimensions after the dimension where the start/end coordinates deviate.

In this example, the `(0,0,0)` and `(0,1,0)` start/end coordinates deviate on dimension-id `1`. We will expand all dimensions with an id `>= 1`, which will just be dimension-id `2`.

This path has been updated to reflect these changes.

Reply:
Indeed, good observation! It looks good now, thanks.